### PR TITLE
fix(upgrade) : upgrade to hapi 17

### PR DESCRIPTION
**Pull request by deeptibaghel**
@vbudhram i am not able to upgrade this plugin to hapi 17, need your help please.

### CODE REVIEW

**Comment by deeptibaghel** [comment]
Here is the error https://pastebin.com/UNVt6QxJ

**Comment by vladikoff** [comment]
We figured out the `TypeError: server.inject is not a function` errors. 
@deeptibaghel let us know what others you are getting

(We should also add some CI to this repo later on)

**Comment by deeptibaghel** [comment]
@vladikoff now it is :
1) HPKP Headers
       "before all" hook:
     Error: listen EADDRINUSE 0.0.0.0:3000
      at Object._errnoException (util.js:1022:11)
      at _exceptionWithHostPort (util.js:1044:20)
      at Server.setupListenHandle [as _listen2] (net.js:1367:14)
      at listenInCluster (net.js:1408:12)
      at doListen (net.js:1517:7)
      at _combinedTickCallback (internal/process/next_tick.js:141:11)
      at process._tickCallback (internal/process/next_tick.js:180:9)

  2) HPKP Headers
       "after all" hook:
     TypeError: Cannot read property 'stop' of undefined
      at Context.<anonymous> (tests/index.js:92:21)

**Comment by vladikoff** [comment]
@deeptibaghel quick tip, for future work make sure to create feature branches for your changes:

image

so instead of `master`, it can be something like `hapi17-update`. This is useful to keep organized

**Comment by deeptibaghel** [comment]
@vladikoff yeah i will take care, i do generally for other repos , will do for all now.

**Comment by vbudhram** [comment]
@deeptibaghel Looks like there are a few things that need to be updated to get this plugin hapi 17 compatible.

From https://tech.xogrp.com/migrating-to-hapi-v17-plugin-structure-and-glue-manifests-a7b3c3d551f2, plugins are registered differently so we will need to update this to new format.

We will also have to update the test server as well.

**Comment by vladikoff** [comment]
some updates listed here: https://github.com/vbudhram/hapi-hpkp/pull/3/files

File: tests/index.js [hunk 1]
```diff
@@ -86,17 +84,14 @@ describe('HPKP Headers', function () {
...
 
-    after(function () {
-      return server.stop()
-    })
-    it(testCase.name, function (done) {
-      server.inject(requestOptions, function (response) {
-        assert.equal(response.headers[testCase.expectedKey], testCase.expectedHeader)
-        done()
+    it(testCase.name, function () {
+      return createServer(3000, testCase.options).then((s) => {
```
Comment (vladikoff) [inline tests/index.js]: @deeptibaghel before merging this let's clean this up to maybe use `before/after` again. More importantly let's see if this plugin works in the fxa-auth-server before merging :)

**Comment by vbudhram** [comment]
@deeptibaghel I updated some of the tests that were causing port conflicts, rebase when you get a chance and see if they are still failing for you.

File: package.json [hunk 2]
```diff
@@ -22,17 +22,18 @@
...
-    "joi": "9.0.4"
+    "joi": "13.3.0"
   },
   "devDependencies": {
-    "chai": "3.5.0",
+    "chai": "4.1.2",
     "eslint-config-fxa": "2.1.0",
-    "hapi": "15.0.2",
+    "eslint-plugin-fxa": "git+https://github.com/mozilla/eslint-plugin-fxa#master",
+    "hapi": "17.4.0",
```
Comment (vladikoff) [inline package.json]: let's update to latest hapi version here

File: tests/index.js [hunk 3]
```diff
@@ -3,31 +3,29 @@ var hpkp = require('../lib/hpkp')
...
-    if (err) {
-      console.error('Failed to load plugin:', err)
-    }
-  });
+  try{
+   await server.register({
+      plugin: require('../index.js'),
+      options: hpkpOptions
+    })
+  }catch(err){
```
Comment (vladikoff) [inline tests/index.js]: spacse between `} catch(err) {`

File: tests/index.js [hunk 4]
```diff
@@ -3,31 +3,29 @@ var hpkp = require('../lib/hpkp')
...
+    console.error('Failed to load plugin:', err)
+  }
 
-  server.route({
+  await server.route({
     method: 'GET',
     path: '/',
     handler: function (request, reply) {
```
Comment (vladikoff) [inline tests/index.js]: should the wait be before `server.route` or before the function here in the handler? `handler: await function`

File: tests/index.js [hunk 5]
```diff
@@ -3,31 +3,29 @@ var hpkp = require('../lib/hpkp')
...
-      return reply('HPKP!')
+      return 'HPKP!'
     }
   })
 
-  server.start()
-
+  await server.start()
```
Comment (vladikoff) [inline tests/index.js]: 👍

File: tests/index.js (hunk 1 above)
Comment (vladikoff) [reply tests/index.js]: Still applies, let's clean this up

**Review by vladikoff — CHANGES_REQUESTED** [review]
Need some changes and a rebase

**Pull request closed without merging**